base/task.jl

```diff
@@ -184,6 +184,7 @@ end
 
 # NOTE: you can only wait for scheduled tasks
 function wait(t::Task)
+    t === current_task() && error("deadlock detected")
```

Comment:
```suggestion
    t === current_task() && error("Possible deadlock detected: Cannot wait on current task")
```

Reply:
Since it's certain deadlock, saying "possible" seems odd

Reply:
I was thinking about user-error, where someone literally does `wait(current_task())`. But in the end I only care that it is clear which action is not allowed.

Reply:
Yes, I ran into this because I did `foreach(wait, tasks)`, then corrupted a huge amount of memory (basically, I just dropped all GC roots) which put current_task into tasks.

In full assertion mode, we could even expand this to check that `donenotify` forms a DAG, this just constitutes a simple (e.g. inexpensive) subset of the general case. I think other systems do something similar to that, but on a timer?